File: results/1.0/HTTP-1.1/http-1.1-claude-3-5-haiku-20241022.c
#include <hammer/hammer.h>
#include <hammer/glue.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

HParser* http_request;
HParser* http_response;
HParser* http_message;

HParser* sp;
HParser* crlf;
HParser* token;
HParser* quoted_string;
HParser* method;
HParser* http_version;
HParser* status_code;

HParser* create_http_parsers() {
    // Basic character parsers
    const uint8_t sp_chars[] = {' '};
    const uint8_t token_exclude[] = {' ', '\r', '\n', '\t', '(', ')', '{', '}', '[', ']', '<', '>', ':', ';', ',', '"', '/', '?', '='};
    const uint8_t quote_exclude[] = {'"'};
    const uint8_t crlf_chars[] = {'\r', '\n'};

    sp = h_ch(sp_chars[0]);
    crlf = h_sequence(h_ch('\r'), h_ch('\n'), NULL);
    token = h_many1(h_not_in(token_exclude, sizeof(token_exclude)));
    quoted_string = h_sequence(
        h_ch('"'),
        h_many(h_not_in(quote_exclude, sizeof(quote_exclude))),
        h_ch('"'),
        NULL
    );

    // HTTP Method parser
    method = h_choice(
        h_parse_string("GET"),
        h_parse_string("POST"),
        h_parse_string("PUT"),
        h_parse_string("DELETE"),
        h_parse_string("HEAD"),
        h_parse_string("OPTIONS"),
        h_parse_string("TRACE"),
        h_parse_string("CONNECT"),
        NULL
    );

    // HTTP Version parser
    http_version = h_sequence(
        h_parse_string("HTTP/"),
        h_choice(
            h_parse_string("1.0"),
            h_parse_string("1.1"),
            NULL
        ),
        NULL
    );

    // Status Code parser
    HParser* zero = h_ch('0');
    HParser* nine = h_ch('9');
    HParser* digit = h_in_range(zero, nine);
    status_code = h_sequence(
        digit, digit, digit,
        NULL
    );

    // Headers parser
    HParser* header_name = token;
    HParser* header_value = h_many1(h_not_in(crlf_chars, sizeof(crlf_chars)));
    
    HParser* header = h_sequence(
        header_name,
        h_ch(':'),
        sp,
        header_value,
        crlf,
        NULL
    );

    HParser* headers = h_many(header);

    // Request Line parser
    HParser* request_line = h_sequence(
        method, sp,
        token, sp,  // Request URI
        http_version,
        crlf,
        NULL
    );

    // Response Line parser
    HParser* response_line = h_sequence(
        http_version, sp,
        status_code, sp,
        h_many(h_not_in(crlf_chars, sizeof(crlf_chars))),  // Reason Phrase
        crlf,
        NULL
    );

    // Message Body parser
    const uint8_t null_char[] = {0};
    HParser* message_body = h_many(h_not_in(null_char, sizeof(null_char)));

    // Complete Request parser
    http_request = h_sequence(
        request_line,
        headers,
        crlf,
        h_optional(message_body),
        NULL
    );

    // Complete Response parser
    http_response = h_sequence(
        response_line,
        headers,
        crlf,
        h_optional(message_body),
        NULL
    );

    // Generic Message parser
    http_message = h_choice(
        http_request,
        http_response,
        NULL
    );

    return http_message;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <http_file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t *buffer = malloc(file_size);
    if (!buffer) {
        perror("Memory allocation error");
        fclose(file);
        return 1;
    }

    size_t read_size = fread(buffer, 1, file_size, file);
    fclose(file);

    if (read_size != file_size) {
        perror("File read error");
        free(buffer);
        return 1;
    }

    create_http_parsers();
    HParseResult* result = h_parse(http_message, buffer, read_size);

    if (result && result->ast) {
        printf("Parsing successful!\n");
        h_parse_result_free(result);
    } else {
        printf("Parsing failed.\n");
    }

    free(buffer);
    return 0;
}

File: results/1.0/HTTP-1.1/http-1.1-claude-3-5-haiku-20241022.rs
use nom::{
    IResult,
    bytes::complete::{tag, take_while, take_while1},
    sequence::{tuple, terminated},
    character::complete::{space1, crlf},
    multi::many0,
    combinator::opt,
};
use std::str::from_utf8;
use std::env;
use std::fs;

#[derive(Debug)]
struct HttpRequest<'a> {
    method: &'a str,
    uri: &'a str,
    version: &'a str,
    headers: Vec<Header<'a>>,
    body: Option<&'a [u8]>,
}

#[derive(Debug)]
struct Header<'a> {
    name: &'a str,
    value: &'a str,
}

fn is_token_char(c: u8) -> bool {
    matches!(c as char, 'a'..='z' | 'A'..='Z' | '0'..='9' | '-' | '_' | '.' | '~')
}

fn parse_method(input: &[u8]) -> IResult<&[u8], &str> {
    let methods = &["GET", "POST", "PUT", "DELETE", "HEAD", "OPTIONS", "TRACE", "CONNECT"];
    let (remaining, method) = take_while1(|c: u8| (c as char).is_ascii_uppercase())(input)?;
    let method_str = from_utf8(method).unwrap();
    
    if methods.contains(&method_str) {
        Ok((remaining, method_str))
    } else {
        Err(nom::Err::Error(nom::error::Error::new(input, nom::error::ErrorKind::Tag)))
    }
}

fn parse_uri(input: &[u8]) -> IResult<&[u8], &str> {
    let (remaining, uri) = take_while1(|c: u8| c != b' ')(input)?;
    Ok((remaining, from_utf8(uri).unwrap()))
}

fn parse_http_version(input: &[u8]) -> IResult<&[u8], &str> {
    let (remaining, version) = terminated(
        tag("HTTP/1.1"),
        crlf
    )(input)?;
    Ok((remaining, "HTTP/1.1"))
}

fn parse_header_name(input: &[u8]) -> IResult<&[u8], &str> {
    let (remaining, name) = take_while1(is_token_char)(input)?;
    Ok((remaining, from_utf8(name).unwrap()))
}

fn parse_header_value(input: &[u8]) -> IResult<&[u8], &str> {
    let (remaining, value) = take_while(|c: u8| c != b'\r')(input)?;
    Ok((remaining, from_utf8(value).unwrap().trim()))
}

fn parse_header(input: &[u8]) -> IResult<&[u8], Header> {
    let (remaining, (name, _, value, _)) = tuple((
        parse_header_name,
        tag(": "),
        parse_header_value,
        crlf
    ))(input)?;

    Ok((remaining, Header { name, value }))
}

fn parse_headers(input: &[u8]) -> IResult<&[u8], Vec<Header>> {
    many0(parse_header)(input)
}

fn parse_body(input: &[u8]) -> IResult<&[u8], Option<&[u8]>> {
    opt(take_while(|_| true))(input)
}

fn parse_http_request(input: &[u8]) -> IResult<&[u8], HttpRequest> {
    let (remaining, (method, _, uri, _, _version)) = tuple((
        parse_method,
        space1,
        parse_uri,
        space1,
        parse_http_version
    ))(input)?;

    let (remaining, headers) = parse_headers(remaining)?;
    let (remaining, body) = parse_body(remaining)?;

    Ok((
        remaining, 
        HttpRequest {
            method,
            uri,
            version: "HTTP/1.1",
            headers,
            body
        }
    ))
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <input_file>", args[0]);
        std::process::exit(1);
    }

    let input_file = &args[1];
    let input_data = fs::read(input_file)?;

    match parse_http_request(&input_data) {
        Ok((_, request)) => {
            println!("{:#?}", request);
            Ok(())
        }
        Err(e) => {
            eprintln!("Parsing error: {:?}", e);
            std::process::exit(1)
        }
    }
}